ethereum/p2p/src/main/java/tech/pegasys/pantheon/ethereum/p2p/network/NodePermissioningAdapter.java

```diff
@@ -0,0 +1,130 @@
+import org.apache.logging.log4j.Logger;
+
+class NodePermissioningAdapter extends PeerPermissions {
+  private static final Logger LOG = LogManager.getLogger();
+
+  private final NodePermissioningController nodePermissioningController;
+  private final List<EnodeURL> bootnodes;
+  private final Blockchain blockchain;
+  private final long blockchainListenId;
+  private final long nodePermissioningListenId;
+
+  public NodePermissioningAdapter(
+      final NodePermissioningController nodePermissioningController,
+      final List<EnodeURL> bootnodes,
+      final Blockchain blockchain) {
+    checkNotNull(nodePermissioningController);
+    checkNotNull(bootnodes);
+    checkNotNull(blockchain);
+
+    this.nodePermissioningController = nodePermissioningController;
+    this.bootnodes = bootnodes;
+    this.blockchain = blockchain;
+
+    // TODO: These events should be more targeted
+    blockchainListenId =
+        blockchain.observeBlockAdded((evt, chain) -> dispatchUpdate(true, Optional.empty()));
+    nodePermissioningListenId =
+        this.nodePermissioningController.subscribeToUpdates(
+            () -> dispatchUpdate(true, Optional.empty()));
+  }
+
+  @Override
+  public boolean isPermitted(final Peer localNode, final Peer remotePeer, final Action action) {
+    switch (action) {
+      case DISCOVERY_ALLOW_IN_PEER_TABLE:
+        return outboundIsPermitted(localNode, remotePeer);
+      case DISCOVERY_ALLOW_OUTBOUND_NEIGHBORS_REQUEST:
+        return allowOutboundNeighborsRequests(localNode, remotePeer);
+      case DISCOVERY_ALLOW_OUTBOUND_BONDING:
+        return allowOutboundBonding(localNode, remotePeer);
+      case DISCOVERY_ACCEPT_INBOUND_BONDING:
+      case DISCOVERY_SERVE_INBOUND_NEIGHBORS_REQUEST:
+      case RLPX_ALLOW_NEW_INBOUND_CONNECTION:
+        return inboundIsPermitted(localNode, remotePeer);
+      case RLPX_ALLOW_NEW_OUTBOUND_CONNECTION:
+        return outboundIsPermitted(localNode, remotePeer);
+      case RLPX_ALLOW_ONGOING_CONNECTION:
+        // TODO: This should probably check outbound || inbound, or the check should be more
+        // granular
+        return outboundIsPermitted(localNode, remotePeer);
+      default:
+        // Return false for unknown / unhandled permissions
+        LOG.error(
+            "Permissions denied for unknown action {}",
+            action.name(),
+            new IllegalStateException("Unhandled permissions action " + action.name()));
+        return false;
+    }
+  }
+
+  private boolean allowOutboundBonding(final Peer localNode, final Peer remotePeer) {
+    boolean outboundMessagingAllowed = outboundIsPermitted(localNode, remotePeer);
+    if (!nodePermissioningController.getSyncStatusNodePermissioningProvider().isPresent()) {
+      return outboundMessagingAllowed;
+    }
+
+    // We're using smart-contract based permissioning
+    // If we're out of sync, only allow bonding to our bootnodes
+    final SyncStatusNodePermissioningProvider syncStatus =
```

Comment:
All node permissioning providers are supposed to be under the `NodePermissioningController` object. The controller will take care of permissioning rules and the precedence order of them. We shouldn't be using it outside the `NodePermissioningController`.

Any reason for using it here?

Reply:
I was trying to account for this code that's currently in the `PeerDiscoveryController`: https://github.com/PegaSysEng/pantheon/blob/8e27ca73d35d9d46b9eb9c549ccb78cfdd8d6fbe/ethereum/p2p/src/main/java/tech/pegasys/pantheon/ethereum/p2p/discovery/internal/PeerDiscoveryController.java#L222-L238

Reply:
Which is removed here: https://github.com/PegaSysEng/pantheon/pull/1471/files#diff-11104ef3adb5bd629a56cc0044b0cc1aL222

Reply:
Also - your comment reminded me to remove the now-dead code here: https://github.com/PegaSysEng/pantheon/pull/1471/commits/af58dba7a6ddf0da6cc95c71c407e7269f14c8ef